cronjob_scripts/trigger_package_build.py

```diff
@@ -0,0 +1,229 @@
+import time
+from typing import List, cast
+
+from cronjob_scripts.types import CrateAndMaybeVersion, GithubAsset
+from cronjob_scripts.architectures import get_build_os, get_target_architectures
+from cronjob_scripts.checkout_worktree import checkout_worktree_for_target
+from cronjob_scripts.get_latest_version import CrateVersionDict, get_latest_version
+from cronjob_scripts.stats import get_requested_crates
+from cronjob_scripts.crates_io_popular_crates import get_crates_io_popular_crates
+
+
+"""
+* construct the following list of crates (n lists for each target):
+  * for each target:
+    * the list of install requests with the "built-from-source" status in the last day (ordered by popularity, including versions - WARNING: this is very likely to result in the same crate being built many times in parallel if someone runs the cronjob manually on ci/locally)
+    * the list of install requests with the "built-from-source" status in the last day (shuffled, including versions)
+    * the list of install requests with any other status in the last day (shuffled, with versions stripped)
+    * popular-crates.txt (shuffled)
+    * a list with just cargo-quickinstall and nothing else
+* shuffle the list of lists and then round-robin between each list:
+  * take the head of the list
+  * if we already checked this package for this target, skip it
+  * if we already tried to build this package 5 times in the last 7 days, skip it (TODO: also record the version in the failure log?)
+  * check its latest version + if we have a package built for it. If we don't have a package built for it, trigger a build
+    * if we already triggered m package builds since the start of the run, exit (probably set this to n times the number of targets?)
+* if we hit a github rate limit when doing this, this is fine and we just bomb out having made some progress?
+* if we get to the end without triggering any builds then that's fine too: there's nothing to do.
+
+"""
+
+
+def main():
+    import sys
+
+    if len(sys.argv) > 1:
+        print(
+            "Usage: INFLUXDB_TOKEN= target= CRATE_CHECK_LIMIT= RECHECK= GITHUB_REPOSITORY= trigger-package-build.py"
+        )
+        if sys.argv[1] == "--help":
+            exit(0)
+        exit(1)
+
+    targets = get_target_architectures()
+    # FIXME: make a better type than tuple here.
+    # We mean list of (type, target, list of crates and versions)
+    queues: list[tuple[str, str, list[CrateAndMaybeVersion]]] = []
+
+    # we shuffle the list of popular crates from crates.io once, and use it for every arch, to slightly
+    # reduce the number of requests we need to make to crates.io. It probably isn't worth it though.
+    popular_crates = get_crates_io_popular_crates()
+    random.shuffle(popular_crates)
+
+    for target in targets:
+        queue = get_requested_crates(period="1 day", target=target)
+        random.shuffle(queue)
+        queues.append(("requested", target, cast(List[CrateAndMaybeVersion], queue)))
```

Comment:
This cast is kind-of horrible. I forgot that list[A] is not a subtype of list[B] even if A is a subtype of B (contravariance or something).

Even more ugly because cast() just ignores type of input.

I don't really care enough though.